cpp/src/arrow/util/async_generator.h

```diff
@@ -810,10 +810,29 @@ class ReadaheadGenerator {
     Future<> final_future = Future<>::Make();
     std::atomic<int> num_running{0};
     std::atomic<bool> finished{false};
-    std::queue<Future<T>> readahead_queue;
+    std::deque<Future<T>> readahead_queue;
+  };
+
+  struct Cleanup {
+    explicit Cleanup(State* state) : state(state) {}
+    ~Cleanup() {
+      // Lose our own consumer reference to the wrapped generator.
+      state->source_generator = {};
+      // If the generator was destroyed before finishing, wait for all running
+      // tasks to end. While it adds latency, it also reduces the risk of
+      // shenanigans happening at process shutdown (e.g. GH-43604).
+      for (const auto& fut : state->readahead_queue) {
+        fut.Wait();
```

Comment:
Unfortunately, this seems to hang when a MergedGenerator is wrapped in ReadaheadGenerator and an error is propagated along.

(for example in `test_dataset.py::test_checksum_write_dataset_read_dataset_to_table`)

@westonpace Thoughts?